### PR TITLE
Feature/tt 3421 sdk core xdc

### PR DESCRIPTION
**Pull request by opefago**

### CODE REVIEW

File: examples/xdc-example/src/app/xdc.log.example.ts [hunk 1]
```diff
@@ -0,0 +1,13 @@
+import { TatumXdcSDK } from '@tatumio/xdc'
+
+const xdcSDK = TatumXdcSDK({ apiKey: '75ea3138-d0a1-47df-932e-acb3ee807dab' })
+
+export async function xdcLogRecordExample() {
```
Comment (samuelsramko) [inline examples/xdc-example/src/app/xdc.log.example.ts:5]: this example is not needed - delete pls

File: examples/xdc-example/src/app/xdc.root.example.ts [hunk 2]
```diff
@@ -0,0 +1,9 @@
+import { Fiat } from '@tatumio/api-client'
+import { TatumXdcSDK } from '@tatumio/xdc'
+
+const xdcSDK = TatumXdcSDK({ apiKey: '75ea3138-d0a1-47df-932e-acb3ee807dab' })
+
+export async function xdcChangeRateExample() {
+  const rate = await xdcSDK.getExchangeRate()
+  const rateWithBasePair = await xdcSDK.getExchangeRate(Fiat.EUR)
```
Comment (samuelsramko) [inline examples/xdc-example/src/app/xdc.root.example.ts:8]: not needed - delete this example pls

File: examples/xdc-example/src/app/xdc.subscriptions.example.ts [hunk 3]
```diff
@@ -0,0 +1,22 @@
+import { TatumXdcSDK } from '@tatumio/xdc'
+
+const xdcSDK = TatumXdcSDK({ apiKey: '75ea3138-d0a1-47df-932e-acb3ee807dab' })
+
+export async function xdcSubscriptionsExample() {
+  const id = await xdcSDK.subscriptions.createSubscription({
+    type: 'ACCOUNT_INCOMING_BLOCKCHAIN_TRANSACTION',
+    attr: {
+      id: '5e6be8e9e6aa436299950c41',
+      url: 'https://webhook.tatum.io/account',
```
Comment (samuelsramko) [inline examples/xdc-example/src/app/xdc.subscriptions.example.ts:10]: this URL is not good - use dashboard.tatum.io/webhook-handler pls

File: examples/xdc-example/src/app/xdc.subscriptions.example.ts [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+    },
+  })
+
+  await xdcSDK.subscriptions.deleteSubscription('5e68c66581f2ee32bc354087')
+  await xdcSDK.subscriptions.disableWebHookHmac()
+  await xdcSDK.subscriptions.enableWebHookHmac({
```
Comment (samuelsramko) [inline examples/xdc-example/src/app/xdc.subscriptions.example.ts:16]: add comment what is HMAC, link to the proper api calls in the API doc for all examples

File: examples/xdc-example/src/app/xdc.tx.example.ts [hunk 5]
```diff
@@ -0,0 +1,309 @@
...
+    signatureId: 'cac88687-33ed-4ca1-b1fc-b02986a90975',
+    nonce: 3252345722143,
+    fee: {
+      gasLimit: '53632',
+      gasPrice: '20',
+    },
+  })
+
+  // ERC20(FUNGIBLE TOKEN)
+  const preparedDeployErc20Transaction = await xdcSDK.erc20.prepare.deploySignedTransaction({
```
Comment (samuelsramko) [inline examples/xdc-example/src/app/xdc.tx.example.ts:30]: dont use prepare functions in examples, use only send - prepare is for our internal use

File: examples/xdc-example/src/app/xdc.tx.example.ts [hunk 6]
```diff
@@ -0,0 +1,309 @@
...
+      recipient: [
+        '0x8cb76aEd9C5e336ef961265c6079C14e9cD3D2eA',
+        '0x8cb76aEd9C5e336ef961265c6079C14e9cD3D2eA',
+        '0x8cb76aEd9C5e336ef961265c6079C14e9cD3D2eA',
+        '0x8cb76aEd9C5e336ef961265c6079C14e9cD3D2eA',
+      ],
+      amount: ['1', '0', '1', '3'],
+    })
+
+  const sentBatchTransferFromCustodialWallet = await xdcSDK.custodial.send.batchTransferFromCustodialWallet({
```
Comment (samuelsramko) [inline examples/xdc-example/src/app/xdc.tx.example.ts:247]: split gas pump to its separate file - so the logic will be in 1 place

File: packages/blockchain/xdc/src/lib/xdc.sdk.ts [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+export const TatumXdcSDK = (args: SDKArguments) => {
+  const web3 = xdcWeb3({ blockchain })
+  const api = XinFinService
+  const txService = xdcTxService({ blockchain, web3 })
+  const { ...evmSdk } = evmBasedSdk({ ...args, blockchain, web3 })
+
+  return {
+    ...evmSdk,
+    kms: xdcKmsService({ blockchain, web3 }),
+    transaction: txService.native,
```
Comment (samuelsramko) [inline packages/blockchain/xdc/src/lib/xdc.sdk.ts:24]: dont you miss virtual accounts and ledger here?

**Review by samuelsramko — APPROVED** [review]

File: packages/shared/blockchain/evm-based/src/lib/evm-based.utils.ts [hunk 8]
```diff
@@ -86,6 +89,30 @@ export const evmBasedUtils = {
     return signedTransaction.rawTransaction
   },
 
+  validateSenderBalance: async (client: Web3, privateKey: string, tx: TransactionConfig) => {
```
Comment (samuelsramko) [inline packages/shared/blockchain/evm-based/src/lib/evm-based.utils.ts:92]: this works for One and Klaytn and XDC? All 3 has different client libs then web3

File: packages/shared/blockchain/evm-based/src/lib/evm-based.utils.ts [hunk 9]
```diff
@@ -86,6 +89,30 @@ export const evmBasedUtils = {
...
+    const { gas, gasPrice, nonce, ...txWithoutGas } = tx
+    let estimate: BigNumber
+    try {
+      estimate = new BigNumber(await client.eth.estimateGas(txWithoutGas))
+    } catch (e) {
+      if (gas) estimate = new BigNumber(gas as string)
+      else throw new EvmBasedSdkError({ error: e as Error })
```
Comment (samuelsramko) [inline packages/shared/blockchain/evm-based/src/lib/evm-based.utils.ts:99]: you must throw more specific error with error message, so inside core api we willl be able to detect itt properly. something like cannot estimate gas, reason:..., e

**Review by samuelsramko — CHANGES_REQUESTED** [review]

File: packages/shared/blockchain/evm-based/src/lib/evm-based.utils.ts [hunk 10]
```diff
@@ -96,7 +96,7 @@ export const evmBasedUtils = {
       estimate = new BigNumber(await client.eth.estimateGas(txWithoutGas))
     } catch (e) {
       if (gas) estimate = new BigNumber(gas as string)
-      else throw new EvmBasedSdkError({ error: e as Error })
+      else throw new EvmBasedSdkError({ error: new Error(`Unable to estimate gas, reason ${e}`) })
```
Comment (samuelsramko) [inline packages/shared/blockchain/evm-based/src/lib/evm-based.utils.ts:99]: not like this, check how we do it e.g. here - https://github.com/tatumio/tatum-js/blob/master/packages/blockchain/xrp/src/lib/services/xrp.tx.ts#L108

File: packages/shared/blockchain/evm-based/src/lib/evm-based.utils.ts (hunk 8 above)
Comment (opefago) [reply packages/shared/blockchain/evm-based/src/lib/evm-based.utils.ts:92]: It is verified from the e2e tests to work for the chains

File: packages/shared/blockchain/evm-based/src/lib/evm-based.utils.ts (hunk 10 above)
Comment (opefago) [reply packages/shared/blockchain/evm-based/src/lib/evm-based.utils.ts:99]: This error is only returned when it fails to calculate gas for some reasons, maybe there is an issue with the transaction data for similar error as shown in the [link]( https://github.com/tatumio/tatum-js/pull/617/files#diff-e0977f8da3d51fe35030e19de0961c735cece95d55bdd64e0c913aeaed337f86R107-R112). What do you think?

Comment (samuelsramko) [reply packages/shared/blockchain/evm-based/src/lib/evm-based.utils.ts:99]: I am just saying how the format of thrown error should look like.

**Review by samuelsramko — APPROVED** [review]

**Pull request merged**